Add the providers information in the email.

awais786 opened the pull request:
ECOM-2931
**Calling ecommerce service to get the credit providers.**
**Added creditconfig table/migrtion to manage the cache_ttl.**
**Pass the information to the email template (HTML/TEXT).**
**Modify the ecom-api-client to handle worker users**
Adding information into the template needs @marcotuts input. Also If api fails to get providers information it will send the old text.

awais786 commented:
@zubair-arbi please review this.

ahsan-ul-haq commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -26,10 +26,13 @@
 from edxmako.template import Template
 from microsite_configuration import microsite
 from xmodule.modulestore.django import modulestore
+from openedx.core.djangoapps.commerce.utils import ecommerce_api_client
```
Alphabetize

ahsan-ul-haq commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -26,10 +26,13 @@
 
 log = logging.getLogger(__name__)
 
+CREDIT_PROVIDERS_CACHE_KEY = 'credit.providers.api.data'
```
I don't think it should be here as I guess we would need to show them on the dashboard as well.

ahsan-ul-haq commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +186,32 @@ def _email_url_parser(url_name, extra_param=None):
     return urlparse.urlunparse(dashboard_link_parts)
+
+
+def _get_credit_providers(user, course_key):
```
This is a utility method would be helpful in future as well so move it https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/credit/utils.py

zubair-arbi replied:
I would suggest moving this method to `openedx/core/djangoapps/commerce/utils.py` and changing its name to `get_credit_providers`.

awais786 replied:
I have moved this into  https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/credit/utils.py

ahsan-ul-haq commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -9,13 +10,16 @@
 from xmodule.modulestore.tests.factories import CourseFactory
 
 from util.date_utils import from_timestamp
+
```
remove this.

zubair-arbi replied:
remove it

ahsan-ul-haq commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -31,9 +35,12 @@
     CreditRequirementStatus,
     CreditEligibility
 )
+from lms.djangoapps.commerce.tests import TEST_API_URL, TEST_API_SIGNING_KEY
```
Alphabetize

ahsan-ul-haq commented on lms/templates/credit_notifications/credit_eligibility_email.html:
```diff
@@ -34,7 +34,17 @@
+                       % endfor
+                     </ul>
+                % endif
+               ${_(u" for successfully completing your {platform_name} course! {link_start}Get your credit now.{link_end}").format(
```
I think `{link_start}Get your credit now.{link_end}` would be different for each provider. So I guess we would need to have a different text here.

awais786 replied:
No it will redirect toward the dashboard.

marcotuts replied:
No we have a provider URL that should be where this goes. Is this not what it does today? I could be wrong.

awais786 replied:
```
<p style="padding:5px 0">
              Congratulations! You are eligible to receive course credit for successfully completing your Your Platform Name Here course! <a href="https://ahsan-ul-haq.sandbox.edx.org/dashboard" target="_blank">Get your credit now.</a>
</p>
```
I have one email so its shows the dashboard. I will verify this again in code too.

ahsan-ul-haq commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +186,32 @@ def _email_url_parser(url_name, extra_param=None):
+                ]
+            )
+
+        cache.set(cache_key, providers_data)
```
Don't we need `cache ttl` here?

ahsan-ul-haq commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +186,32 @@ def _email_url_parser(url_name, extra_param=None):
+
+    if not providers_data:
+        providers_data = []
+        response = ecommerce_api_client(user).courses(unicode(course_key) + '/?include_products=1').get()
```
Why are we adding `/?include_products=1`?

awais786 replied:
It will return the complete information of a course.

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -92,6 +99,45 @@ class CreditRequirementApiTests(CreditApiTestBase):
+        """ Mock GET requests to the ecommerce course API endpoint. """
+        httpretty.reset()
+        httpretty.register_uri(
+            httpretty.GET, "{}/courses/{}/?include_products=1".format(TEST_API_URL, unicode(self.course_key)),
```
use single quotes

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -92,6 +99,45 @@ class CreditRequirementApiTests(CreditApiTestBase):
+        )
+
+    @httpretty.activate
+    @override_settings(ECOMMERCE_API_URL=TEST_API_URL, ECOMMERCE_API_SIGNING_KEY=TEST_API_SIGNING_KEY)
```
move this settings override to class level as I am seeing it on almost all tests

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -92,6 +99,45 @@ class CreditRequirementApiTests(CreditApiTestBase):
+    @httpretty.activate
+    @override_settings(ECOMMERCE_API_URL=TEST_API_URL, ECOMMERCE_API_SIGNING_KEY=TEST_API_SIGNING_KEY)
+    def test_get_credit_providers_caching(self):
+        """Verify that providers list is cached for 5 minutes."""
```
```
"""Verify that providers list is cached."""
```

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -549,6 +601,112 @@ def test_set_credit_requirement_status_req_not_configured(self):
         self.assertEqual(len(req_status), 1)
         self.assertEqual(req_status[0]["status"], None)
 
+    def course_api_response_data(self):
```
make it a test mixin

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -549,6 +601,112 @@ def test_set_credit_requirement_status_req_not_configured(self):
+        """ecommerce course api response data."""
+
+        return {
+            "id": "course-v1:arb+cs999+2015_t2",
```
use demo course id as an example

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -549,6 +601,112 @@ def test_set_credit_requirement_status_req_not_configured(self):
+
+        return {
+            "id": "course-v1:arb+cs999+2015_t2",
+            "url": "http://localhost:8002/api/v2/courses/course-v1:arb+cs999+2015_t2/",
```
be persistent with single quotes

awais786 replied:
In this file double quotes in use every where.

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -9,13 +10,16 @@
 from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
 from xmodule.modulestore.tests.factories import CourseFactory
 
 from util.date_utils import from_timestamp
```
order imports alphabetically

zubair-arbi commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -549,6 +601,112 @@ def test_set_credit_requirement_status_req_not_configured(self):
+            ]
+        }
+
+    def _assert_providers_data(self, html_content_first):
```
add docstring for this helper method

awais786 commented on lms/templates/credit_notifications/credit_eligibility_email.html:
```diff
@@ -34,7 +34,17 @@
             <p>
-              ${_(u"Congratulations! You are eligible to receive course credit for successfully completing your {platform_name} course! {link_start}Get your credit now.{link_end}").format(
+               ${_(u"Congratulations! You are eligible to receive course credit from ")}
+                % if providers:
```
@marcotuts Please guide how to show the multiple providers. simple comma separated list is fine or some thing else ?

marcotuts replied:
Comma separated list is good. {provider-1}, {provider-2}, or {provider3-}

awais786 commented:
@marcotuts also tell me if api fails to get data than how to handle that scenario ?

awais786 commented:
@clintonb , @rlucioni Please review this PR. Few minor things left e.g displaying data into the templates.
But you can do 1st round of review.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -67,6 +68,9 @@ def send_credit_notifications(username, course_key):
         logo_image_id = logo_image.get('Content-ID', '')[1:-1]
 
+    providers = get_credit_providers(user, course_key)
+    providers = ", ".join(providers) if providers else None
```
It looks like this providers list is going to be used in sentences displayed to the user. At the moment, this is just going to return a comma-separated list of providers. It would be better if the number of providers available influenced the structure of the string:
```
One provider => "A"
Two providers => "A and B"
Three or more providers => "A, B, ..., and N"
```
This is also a tricky i18n problem, but fortunately one that we've solved before: https://github.com/edx/edx-platform/commit/1d1baa50098c1108a4ae7e0c2fc345fbc8dd8a3c#diff-d8e9b05acd3b7dba84709ae01b99053cR858

awais786 commented:
<img width="816" alt="screen shot 2016-03-11 at 7 16 22 pm" src="https://cloud.githubusercontent.com/assets/445320/13704439/ce253a2c-e7bd-11e5-96a9-d6ba3836f862.png">
With 2 providers.

awais786 commented:
<img width="792" alt="screen shot 2016-03-11 at 7 17 46 pm" src="https://cloud.githubusercontent.com/assets/445320/13704469/f8fa8f18-e7bd-11e5-87f3-e49600feee78.png">
with 3 provider

awais786 commented:
<img width="806" alt="screen shot 2016-03-11 at 7 22 26 pm" src="https://cloud.githubusercontent.com/assets/445320/13704612/9e98446a-e7be-11e5-9658-f832195efc33.png">
With 2 providers

awais786 commented:
<img width="799" alt="screen shot 2016-03-11 at 7 27 47 pm" src="https://cloud.githubusercontent.com/assets/445320/13704771/5eb80140-e7bf-11e5-881f-ccec96c33c60.png">
One provider

awais786 commented:
@rlucioni Please review this PR. Also I have implemented the cache for providers data. 
Please let me know if it is not required.

awais786 commented:
@marcotuts I have added screenshots. Please check that.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -67,6 +68,26 @@ def send_credit_notifications(username, course_key):
+    if providers:
+        if len(providers) > 1:
+            if len(providers) > 2:
+                # Translators: The join of three or more providers names (e.g., ASU, MIT, and Charter OAK).
```
"join of three or more *university* names"

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -67,6 +68,26 @@ def send_credit_notifications(username, course_key):
+            if len(providers) > 2:
+                # Translators: The join of three or more providers names (e.g., ASU, MIT, and Charter OAK).
+                providers_string = _("{first_provider}, and {last_provider}").format(
+                    first_provider=u", ".join(providers[:-1]),
```
`first_providers`

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -75,6 +96,7 @@ def send_credit_notifications(username, course_key):
         'dashboard_link': dashboard_link,
         'credit_course_link': credit_course_link,
         'tracking_pixel': tracking_pixel,
+        'providers': providers_string,
```
You should be using ungettext here to account for the fact that languages deal with pluralization differently. See [here](https://github.com/edx/edx-platform/commit/1d1baa50098c1108a4ae7e0c2fc345fbc8dd8a3c#diff-d8e9b05acd3b7dba84709ae01b99053cR874) for an example.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -67,6 +68,26 @@ def send_credit_notifications(username, course_key):
+        if len(providers) > 1:
+            if len(providers) > 2:
+                # Translators: The join of three or more providers names (e.g., ASU, MIT, and Charter OAK).
+                providers_string = _("{first_provider}, and {last_provider}").format(
```
`first_providers`. You need to make it clear to translators that the first of these formatting strings represents a list of nouns.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -85,6 +107,8 @@ def send_credit_notifications(username, course_key):
     # render the credit notification templates
     subject = _(u'Course Credit Eligibility')
+    if providers:
+        subject = _(u'You are eligible for credit from {providers}').format(providers=providers_string)
```
This string should also be the return value of `ungettext`, described above.

awais786 commented:
@rlucioni Please review this PR again.

awais786 commented:
<img width="887" alt="screen shot 2016-03-14 at 6 59 51 pm" src="https://cloud.githubusercontent.com/assets/445320/13746819/fd3d4402-ea16-11e5-8581-d4c96e584be5.png">

Email with display names.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -67,6 +69,26 @@ def send_credit_notifications(username, course_key):
+        if len(providers) > 1:
+            if len(providers) > 2:
+                # Translators: First of these formatting strings represents a list of nouns
+                # university names (e.g., ASU, MIT, and Charter OAK).
```
> # Translators: The join of three or more university names. The first of these formatting strings represents a comma-separated list of names (e.g., MIT, Harvard, Dartmouth).

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -75,6 +97,8 @@ def send_credit_notifications(username, course_key):
         'dashboard_link': dashboard_link,
         'credit_course_link': credit_course_link,
         'tracking_pixel': tracking_pixel,
+        'providers': ungettext('{providers_string}', '{providers_string}', len(providers)).format(
```
I take back my earlier comment about `ungettext`; sorry! `ungettext` takes three arguments: the singular translation string, the plural translation string, and the number of objects. In this case, there's actually not a difference between the singular and plural translation strings. They're both just lists of names, so I don't think there's a need to call `ungettext` here.

See https://docs.djangoproject.com/en/1.9/topics/i18n/translation/#pluralization for more.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -85,6 +109,11 @@ def send_credit_notifications(username, course_key):
     # render the credit notification templates
     subject = _(u'Course Credit Eligibility')
+    if providers:
+        subject = ungettext(
```
Same comment about `ungettext` here. I don't think it's necessary.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +209,50 @@ def _email_url_parser(url_name, extra_param=None):
+
+def get_credit_providers(user, course_key):
+    """Get the course production information from ecommerce and parse the data
+     to get providers.
```
> Get the course information from ecommerce and parse the data to get providers.

rlucioni commented:
:+1: once comments addressed and commits squashed.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +207,49 @@ def _email_url_parser(url_name, extra_param=None):
+            [provider['display_name'] for provider in providers_information if provider['id'] == provider_id]
+        )
+
+    cache.set(cache_key, providers_display_names)
```
@awais786 you're not passing a cache timeout here, which means the given value will be cached forever. For more, see https://docs.djangoproject.com/en/1.9/topics/cache/#basic-usage. Credit providers may not change often, but they will change. This value shouldn't be cached forever.

awais786 replied:
@rlucioni Thanks for the catch. My assumption was it will cache for 5 mins If I pass nothing. 300 seconds fine as cache timeout ?

rlucioni replied:
5 minutes seems reasonable, but you shouldn't hard-code the value. Is there an existing config model you can use?

awais786 replied:
@rlucioni There is existing config available in credit models. Only one thing is available but I don't think it is useful for us.
https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/credit/models.py#L207

We can define the variable
```
CREDIT_PROVIDERS_CACHE_TIMEOUT = 300
```

rlucioni replied:
@awais786 the model you linked to isn't a [ConfigurationModel](https://github.com/edx/edx-platform/blob/master/common/djangoapps/config_models/models.py#L61), which means it won't give us an audit trail or deal with cache clearing when we change values. The value being configured here is used when making requests to the ecommerce service. Would it make sense to use [CommerceConfiguration](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/commerce/models.py#L10)?

awais786 replied:
@rlucioni this model has no cache_ttl value. but If any app can be use than this is more reasonable. [ProgramsConfig](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/programs/models.py#L45)

rlucioni replied:
@awais786 if a field you need is missing, add it. Don't use the ProgramsConfig; this has nothing to do with programs.

awais786 replied:
@rlucioni I think best thing is to add the config model in credit app. It will be useful for future as well. CommerceConfig is giving circular issues at the moment. If they did't resolve I will add the config model in credit app.

awais786 commented:
@marcotuts I have added a screen shot where I am showing the provider display name instead of ids. Please let me know if you want to change it. Otherwise PR is ready to merge.

marcotuts commented:
looks good to me!

zubair-arbi commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,58 @@ def _email_url_parser(url_name, extra_param=None):
+    providers_data = None
+    cache_key = None
+    if credit_config.enabled:
+        cache_key = credit_config.CACHE_KEY + '.' + unicode(course_key)
```
```
cache_key = '{key_prefix}.{course_key}'.format(key_prefix=credit_config.CACHE_KEY course_key=unicode(course_key))
```

zubair-arbi commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -17,12 +17,14 @@
 from model_utils.models import TimeStampedModel
 from xmodule_django.models import CourseKeyField
 from django.utils.translation import ugettext_lazy
+from config_models.models import ConfigurationModel
```
order imports alphabetically

awais786 commented:
@rlucioni Please review the config implementation.

zubair-arbi commented:
@awais786 changes looks good :+1: 
Please make sure that you have added a follow up story/devops ticket to enable CreditApiConfig on prod/stage.

rlucioni commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -709,3 +708,28 @@ def __unicode__(self):
+        """Whether responses from the commerce API will be cached."""
+        return self.cache_ttl > 0
+
+    def __unicode__(self):
```
This isn't necessary.

awais786 replied:
I was getting quality error on jenkins 'no __unicode__ method on model'

rlucioni replied:
It's not necessary. However, if you want to provide a \__unicode\__ representation, make it something useful (e.g., Credit Configuration). This method is used to generate the table name displayed in the Django admin.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -392,10 +487,14 @@ def test_remove_credit_requirement_status_req_not_configured(self):
         """ Test the credit requirements, eligibility notification, email
         content caching for a credit course.
         """
+        self._mock_ecommerce_courses_api(self.course_key, self.COURSE_API_RESPONSE, 200)
```
The default status is 200, no need to provide it here.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -876,3 +1070,126 @@ def _assert_credit_status(self, expected_status):
+    def test_get_credit_providers_method(self):
+        """Verify that parsed providers list is returns after getting course production information."""
+        user = UserFactory.create()
+        self._mock_ecommerce_courses_api(self.course_key, self.COURSE_API_RESPONSE, 200)
```
The default status is 200, no need to provide it here.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -876,3 +1070,126 @@ def _assert_credit_status(self, expected_status):
+    @httpretty.activate
+    def test_get_credit_providers_caching(self):
+        """Verify that providers list is cached."""
+        self._mock_ecommerce_courses_api(self.course_key, self.COURSE_API_RESPONSE, 200)
```
The default status is 200, no need to provide it here.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -876,3 +1070,126 @@ def _assert_credit_status(self, expected_status):
+        response_providers = get_credit_providers_by_course(user, self.course_key)
+        self.assertEqual(self.PROVIDERS_LIST, response_providers)
+
+        # Verify only one request was made.
```
The comments in this method are not accurate. Remove them.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -876,3 +1070,126 @@ def _assert_credit_status(self, expected_status):
+        self.credit_config.cache_ttl = 0
+        self.credit_config.save()
+
+        self._mock_ecommerce_courses_api(self.course_key, self.COURSE_API_RESPONSE, 200)
```
The default status is 200, no need to provide it here.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -876,3 +1070,126 @@ def _assert_credit_status(self, expected_status):
+    @ddt.unpack
+    def test_get_provider_api_with_multiple_data(self, data, expected_data):
+        user = UserFactory.create()
+        self._mock_ecommerce_courses_api(self.course_key, data, 200)
```
The default status is 200, no need to provide it here.

rlucioni commented on openedx/core/djangoapps/credit/tests/test_api.py:
```diff
@@ -876,3 +1070,126 @@ def _assert_credit_status(self, expected_status):
+        self.assertEqual(len(httpretty.httpretty.latest_requests), 1)
+
+        # # Hit the api with clearing cache.
+        for __ in range(3):
```
I don't think this is necessary. The goal of this test is to verify that caching works, which it already has by this point.

rlucioni commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -709,3 +708,28 @@ def __unicode__(self):
         )
+
+
+class CreditApiConfig(ConfigurationModel):
```
Make this more general (`CreditConfig`) and update the docstring to indicate that this is meant to manage credit configuration.

awais786 commented:
@rlucioni all modifications are implemented.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+    credit_config = CreditConfig.current()
+
+    # Bypass caching for staff users, who may be adding providers and want
+    # to see them immediately.
```
This comment isn't accurate. Please remove it. You're doing no such bypassing of the cache here.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+        return
+
+    for product in response.get('products'):
+        providers_data.extend(
```
Just use `+=`.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+
+    providers_display_names = []
+    for provider_id in providers_data:
+        providers_display_names.extend(
```
Just use `+=`.

rlucioni commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -709,3 +708,25 @@ def __unicode__(self):
+    @property
+    def is_cache_enabled(self):
+        """Whether responses from the commerce API will be cached."""
+        return self.cache_ttl > 0
```
This should also check if the config itself is enabled.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+
+    providers_data = None
+    cache_key = None
+    if credit_config.enabled:
```
Use `is_cache_enabled`.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+            [provider['display_name'] for provider in providers_information if provider['id'] == provider_id]
+        )
+
+    if cache_key:
```
Use `is_cache_enabled`. The presence of the cache key is not sufficient when deciding whether to use the cache.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -75,6 +96,7 @@ def send_credit_notifications(username, course_key):
         'dashboard_link': dashboard_link,
         'credit_course_link': credit_course_link,
         'tracking_pixel': tracking_pixel,
+        'providers': providers_string if providers else None,
```
`providers_string or None`

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+        response = ecommerce_api_client(user).courses(unicode(course_key) + '/?include_products=1').get()
+    except Exception:  # pylint: disable=broad-except
+        log.exception("Failed to receive data from the ecommerce course API for Course ID '%s'.", unicode(course_key))
+        return
```
The function docstring claims that a list will be returned. Why not return `providers_data`, which you've set to the empty list above?

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+
+    if not response:
+        log.debug("No Course information found from ecommerce API for Course ID '%s'.", unicode(course_key))
+        return
```
The function docstring claims that a list will be returned. Why not return `providers_data`, which you've set to the empty list above?

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+
+    providers_data = []
+    try:
+        response = ecommerce_api_client(user).courses(unicode(course_key) + '/?include_products=1').get()
```
@awais786 it looks like permissions are going to be a problem here. The ecommerce course API is only usable by staff users.

rlucioni replied:
@awais786 have you looked at [`get_edx_api_data`](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/edx_api_utils.py#L14)? It's meant to be used in situations like these.

awais786 replied:
@rlucioni lms is communicating with ecommerce in a same way. e.g
[checkout page](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/verify_student/views.py#L700)

rlucioni replied:
@awais786 in your example, the LMS is making requests to an endpoint accessible to regular users. In this case, requests are being made to an endpoint which is only accessible to staff users. This is the real problem I'm trying to point out.

[`get_edx_api_data`](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/edx_api_utils.py#L14) is used to communicate with programs and credentials in situations like these. The only obstacle I can think of to using it here is that the ecommerce OAuth client's credentials aren't used to sign JWTs.

awais786 replied:
@rlucioni this method uses [get_id_token](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/token_utils.py#L12). And it uses preferred-username in payload which is different from ecommerece. I have tried to pass username in payload aswell but it fail. One thing I tried also I passed the ECOMMERCE_SERVICE_WORKER_USERNAME user object in [ecommerce_api_client](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/commerce/utils.py#L29) and it works. But for this we need to define its userprofile or modify or create new one.
@clintonb would you please suggest how to access course api in ecommerece. This api need admin permission.

clintonb replied:
Use `ecommerce_api_client`. Update the call to `user.profile.name` to use `None` if the user has n profile.

Also, stop using string concatenation for URLs. See http://slumber.readthedocs.org/en/v0.6.0/tutorial.html#url-parameters.

clintonb commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,60 @@ def _email_url_parser(url_name, extra_param=None):
+        return
+
+    if not response:
+        log.debug("No Course information found from ecommerce API for Course ID '%s'.", unicode(course_key))
```
This seems a bit higher priority than `debug`.

awais786 commented:
@rlucioni I have added a test client we can add new or modify existing to handler users without profile.

awais786 commented:
@rlucioni I have implemented the api client with service user. I have also added new factory to create users without profiles. I can also delete the profile instead of factory. At the moment It has some conflicts in tests cases I will resolve this tomorrow.

awais786 commented:
@rlucioni Please review it.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+    credit_config = CreditConfig.current()
+
+    providers_data = None
+    cache_key = None
```
This isn't necessary.

awais786 replied:
We need to declare it here. Otherwise it will initializes inside the if condition and gives error while setting the cache.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+    if providers_data is not None:
+        return providers_data
+
+    providers_data = []
```
Do this up above instead of setting `providers_data` to `None`.

awais786 replied:
@rlucioni providers_data sets to None if cache value is not available.
```
providers_data = cache.get(cache_key)
```
That's why I initialized it here again.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+        )
+        providers_data = cache.get(cache_key)
+
+    if providers_data is not None:
```
If you make the other suggested changes, this can be simplified to `if providers_data`.

awais786 replied:
@rlucioni If providers_name does not find any provider than it will set the cache with [] for specific time. And if we do following 
```
if providers_data:
```
it will return false again try to hit the ecommerce. 
That's i did this.
```
if providers_data is not None:
```
It will return the True and return the empty list unless cache expire.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+        return providers_data
+
+    if not response:
+        log.info("No Course information found from ecommerce API for Course ID '%s'.", unicode(course_key))
```
You use `unicode(course_key)` three separate times here. Instead, do it once, and assign the return value to `course_id`, conventionally a string.
```python
course_id = unicode(course_key)
```

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+
+    try:
+        user = User.objects.get(username=settings.ECOMMERCE_SERVICE_WORKER_USERNAME)
+        response = ecommerce_api_client(user).courses(unicode(course_key)).get(include_products=1)
```
Nitpick: Since you're using Slumber here, this is going to be deserialized data, not quite a response. For readability, consider renaming this `course_data`.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+            attr.get('value') for attr in product.get('attribute_values') if attr.get('name') == 'credit_provider'
+        ]
+
+    # Hit the CreditProvider method to get the all active providers with complete information.
```
This comment isn't necessary.

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
+    for provider_id in providers_data:
+        providers_display_names += [
+            provider['display_name'] for provider in providers_information if provider['id'] == provider_id
+        ]
```
You can do this more cleanly. For example:
```python
providers = CreditProvider.get_credit_providers()
for provider in providers:
    if provider['id'] in provider_ids:
        provider_display_names += provider['display_name']
```
If `providers_data` contains provider IDs, rename it `provider_ids` for readability. Same with `providers_information`, which contains `providers`.

awais786 replied:
minor change
```
provider_display_names.append(provider['display_name'])
```

rlucioni commented on openedx/core/djangoapps/credit/email_utils.py:
```diff
@@ -180,3 +206,57 @@ def _email_url_parser(url_name, extra_param=None):
     return urlparse.urlunparse(dashboard_link_parts)
+
+
+def get_credit_providers_by_course(course_key):
```
This function returns credit provider display names. Name the function accordingly (e.g., `get_credit_provider_display_names`). Fix the docstring, too; the function returns a list of credit provider display names.

rlucioni commented on lms/djangoapps/commerce/tests/__init__.py:
```diff
@@ -99,3 +100,44 @@ def test_client_unicode(self):
+    @httpretty.activate
+    @freeze_time('2015-7-2')
+    @override_settings(JWT_ISSUER='http://example.com/oauth', JWT_EXPIRATION=30)
+    def test_tracking_context_with_worker_user(self):
```
From the looks of it, you're adding all this to test your change to [API client initialization](https://github.com/edx/edx-platform/pull/11759/files#diff-29385ceb5d39ae786ef9c70a4d7c6b58R35). Instead of copying `test_tracking_context` wholesale, be more precise about what you're testing. All you need to verify is that you can initialize an API client for a user without a profile.

awais786 commented:
@rlucioni Please review this again.

rlucioni commented:
:+1:
